source/common/formatter/substitution_formatter.cc

```diff
@@ -370,18 +370,25 @@ std::vector<FormatterProviderPtr> SubstitutionFormatParser::parse(const std::str
         formatters.push_back(
             std::make_unique<FilterStateFormatter>(key, max_length, serialize_as_string));
       } else if (absl::StartsWith(token, "START_TIME")) {
-        const size_t parameters_length = pos + StartTimeParamStart + 1;
-        const size_t parameters_end = command_end_position - parameters_length;
-
+        const size_t parameters_length = token.length() - (StartTimeParamStart + 1);
         const std::string args = token[StartTimeParamStart - 1] == '('
-                                     ? token.substr(StartTimeParamStart, parameters_end)
+                                     ? token.substr(StartTimeParamStart, parameters_length)
                                      : "";
-        // Validate the input specifier here. The formatted string may be destined for a header, and
-        // should not contain invalid characters {NUL, LR, CF}.
-        if (std::regex_search(args, getStartTimeNewlinePattern())) {
-          throw EnvoyException("Invalid header configuration. Format string contains newline.");
-        }
         formatters.emplace_back(FormatterProviderPtr{new StartTimeFormatter(args)});
+      } else if (absl::StartsWith(token, "DOWNSTREAM_PEER_CERT_V_START")) {
+        const size_t parameters_length = token.length() - (DownstreamPeerCertVStartParamStart + 1);
+        const std::string args =
+            token[DownstreamPeerCertVStartParamStart - 1] == '('
+                ? token.substr(DownstreamPeerCertVStartParamStart, parameters_length)
+                : "";
```

Comment:
Can this part be somehow pushed down into the shared parser/formatter? It's pretty confusing and could use more comments anyway. Until this code is rewritten to use an actual parser grammar I would err on the side of more comments.

Reply:
Sounds right. I'll add a shared format parsing function and push the logic into each constructor, which should allow this to become an easy one liner.